src/undo/EraseUndoAction.cpp

```diff
@@ -1,86 +1,48 @@
 #include "EraseUndoAction.h"
 
-#include "gui/Redrawable.h"
 #include "model/Layer.h"
 #include "model/Stroke.h"
 #include "model/eraser/EraseableStroke.h"
 
-#include "PageLayerPosEntry.h"
 #include "i18n.h"
 
 EraseUndoAction::EraseUndoAction(const PageRef& page): UndoAction("EraseUndoAction") { this->page = page; }
 
-EraseUndoAction::~EraseUndoAction() {
-    for (GList* l = this->original; l != nullptr; l = l->next) {
-        auto* e = static_cast<PageLayerPosEntry<Stroke>*>(l->data);
-        if (!undone) {
-            delete e->element;
-        }
-        delete e;
-    }
-    g_list_free(this->original);
-    this->original = nullptr;
-
-    for (GList* l = this->edited; l != nullptr; l = l->next) {
-        auto* e = static_cast<PageLayerPosEntry<Stroke>*>(l->data);
-        if (undone) {
-            delete e->element;
-        }
-        delete e;
-    }
-    g_list_free(this->edited);
-    this->edited = nullptr;
-}
-
-void EraseUndoAction::addOriginal(Layer* layer, Stroke* element, int pos) {
-    this->original = g_list_insert_sorted(this->original, new PageLayerPosEntry<Stroke>(layer, element, pos),
-                                          reinterpret_cast<GCompareFunc>(PageLayerPosEntry<Stroke>::cmp));
-}
+void EraseUndoAction::addOriginal(Layer* layer, Stroke* element, int pos) { original.emplace(layer, element, pos); }
 
-void EraseUndoAction::addEdited(Layer* layer, Stroke* element, int pos) {
-    this->edited = g_list_insert_sorted(this->edited, new PageLayerPosEntry<Stroke>(layer, element, pos),
-                                        reinterpret_cast<GCompareFunc>(PageLayerPosEntry<Stroke>::cmp));
-}
+void EraseUndoAction::addEdited(Layer* layer, Stroke* element, int pos) { edited.emplace(layer, element, pos); }
 
 void EraseUndoAction::removeEdited(Stroke* element) {
-    for (GList* l = this->edited; l != nullptr; l = l->next) {
-        auto* p = static_cast<PageLayerPosEntry<Stroke>*>(l->data);
-        if (p->element == element) {
-            this->edited = g_list_delete_link(this->edited, l);
-            delete p;
-            p = nullptr;
+    for (auto entryIter = edited.begin(); entryIter != edited.end(); entryIter++) {
+        if (entryIter->element == element) {
+            edited.erase(entryIter);
             return;
         }
     }
 }
 
 void EraseUndoAction::finalize() {
-    for (GList* l = this->original; l != nullptr;) {
-        auto* p = static_cast<PageLayerPosEntry<Stroke>*>(l->data);
-        GList* del = l;
-        l = l->next;
-
-        if (p->element->getPointCount() == 0) {
-            this->edited = g_list_delete_link(this->edited, del);
-            delete p;
-            p = nullptr;
+    for (auto entryIter = original.begin(); entryIter != original.end(); entryIter++) {
```

Comment:
why not using a foreach loop here:

```suggestion
    for (auto&& entry: original) {
```